File: next-flutter/lib/core/storage/document_vault.dart
class VaultException implements Exception {
  VaultException(this.message);
  final String message;
  @override
  String toString() => message;
}

enum VaultSort { name, modified, created }

class VaultFolder {
  VaultFolder(
      {required this.id,
      required this.toolId,
      required this.name,
      this.parentId,
      DateTime? created,
      DateTime? modified})
      : created = created ?? DateTime.now(),
        modified = modified ?? DateTime.now();

  final String id;
  final String toolId;
  String? parentId;
  String name;
  DateTime created;
  DateTime modified;

  Map<String, Object?> toJson() => {
        'id': id,
        'toolId': toolId,
        'parentId': parentId,
        'name': name,
        'created': created.toIso8601String(),
        'modified': modified.toIso8601String(),
      };

  factory VaultFolder.fromJson(Map<String, Object?> json) => VaultFolder(
        id: json['id'] as String,
        toolId: json['toolId'] as String,
        name: json['name'] as String,
        parentId: json['parentId'] as String?,
        created: DateTime.tryParse(json['created'] as String? ?? '') ??
            DateTime.now(),
        modified: DateTime.tryParse(json['modified'] as String? ?? '') ??
            DateTime.now(),
      );
}

class VaultDocument {
  VaultDocument({
    required this.id,
    required this.toolId,
    required this.title,
    this.content = '',
    this.parentId,
    this.query = '',
    this.output = '',
    Map<String, Object?>? metadata,
    DateTime? created,
    DateTime? modified,
  })  : created = created ?? DateTime.now(),
        modified = modified ?? DateTime.now(),
        metadata = metadata ?? {};

  final String id;
  final String toolId;
  String? parentId;
  String title;
  String content;
  String query;
  String output;
  final Map<String, Object?> metadata;
  DateTime created;
  DateTime modified;

  Map<String, Object?> toJson() => {
        'id': id,
        'toolId': toolId,
        'parentId': parentId,
        'title': title,
        'content': content,
        'query': query,
        'output': output,
        'metadata': metadata,
        'created': created.toIso8601String(),
        'modified': modified.toIso8601String(),
      };

  factory VaultDocument.fromJson(Map<String, Object?> json) => VaultDocument(
        id: json['id'] as String,
        toolId: json['toolId'] as String,
        title: json['title'] as String,
        content: json['content'] as String? ?? '',
        parentId: json['parentId'] as String?,
        query: json['query'] as String? ?? '',
        output: json['output'] as String? ?? '',
        metadata: Map<String, Object?>.from(json['metadata'] as Map? ?? {}),
        created: DateTime.tryParse(json['created'] as String? ?? '') ??
            DateTime.now(),
        modified: DateTime.tryParse(json['modified'] as String? ?? '') ??
            DateTime.now(),
      );
}

class VaultPreferences {
  VaultPreferences({
    this.selectedEntryId,
    Set<String>? expanded,
    this.query = '',
    this.includeContent = true,
    this.sort = VaultSort.name,
    this.treeVisible = true,
  }) : expanded = expanded ?? {};

  String? selectedEntryId;
  Set<String> expanded;
  String query;
  bool includeContent;
  VaultSort sort;
  bool treeVisible;

  Map<String, Object?> toJson() => {
        'selectedEntryId': selectedEntryId,
        'expanded': expanded.toList(),
        'query': query,
        'includeContent': includeContent,
        'sort': sort.name,
        'treeVisible': treeVisible,
      };

  factory VaultPreferences.fromJson(Map<String, Object?> json) =>
      VaultPreferences(
        selectedEntryId: json['selectedEntryId'] as String?,
        expanded: {...?json['expanded'] as List?},
        query: json['query'] as String? ?? '',
        includeContent: json['includeContent'] as bool? ?? true,
        sort: VaultSort.values.firstWhere((value) => value.name == json['sort'],
            orElse: () => VaultSort.name),
        treeVisible: json['treeVisible'] as bool? ?? true,
      );
}

class VaultNode {
  VaultNode(
      {required this.id,
      required this.title,
      required this.path,
      required this.isFolder,
      required this.modified,
      required this.created,
      required this.children});
  final String id;
  final String title;
  final String path;
  final bool isFolder;
  final DateTime modified;
  final DateTime created;
  final List<VaultNode> children;
}

class DocumentVault {
  DocumentVault({List<VaultDocument>? documents, List<VaultFolder>? folders})
      : documents = documents ?? [],
        folders = folders ?? [];

  final List<VaultDocument> documents;
  final List<VaultFolder> folders;

  static const supportedTools = ['json', 'quickNote'];

  void validate() {
    final ids = [
      ...documents.map((item) => item.id),
      ...folders.map((item) => item.id)
    ];
    if (ids.toSet().length != ids.length) throw VaultException('文档库包含重复标识。');
    final byId = {for (final folder in folders) folder.id: folder};
    for (final folder in folders) {
      if (!supportedTools.contains(folder.toolId))
        throw VaultException('文件夹所属工具无效。');
      validName(folder.name);
      var visited = {folder.id};
      var parent = folder.parentId;
      while (parent != null) {
        final ancestor = byId[parent];
        if (ancestor == null ||
            ancestor.toolId != folder.toolId ||
            !visited.add(parent) ||
            visited.length > 64) {
          throw VaultException('文件夹层级无效、包含循环或超过 64 层。');
        }
        parent = ancestor.parentId;
      }
    }
    for (final file in documents) {
      if (!supportedTools.contains(file.toolId))
        throw VaultException('文档所属工具无效。');
      if (file.parentId != null && byId[file.parentId]?.toolId != file.toolId) {
        throw VaultException('文档的父文件夹不存在或属于其他工具。');
      }
    }
  }

  String? parentOf(String id) =>
      folders
          .cast<VaultFolder?>()
          .firstWhere((item) => item!.id == id, orElse: () => null)
          ?.parentId ??
      documents
          .cast<VaultDocument?>()
          .firstWhere((item) => item!.id == id, orElse: () => null)
          ?.parentId;

  String? toolOf(String id) =>
      folders
          .cast<VaultFolder?>()
          .firstWhere((item) => item!.id == id, orElse: () => null)
          ?.toolId ??
      documents
          .cast<VaultDocument?>()
          .firstWhere((item) => item!.id == id, orElse: () => null)
          ?.toolId;

  String? nameOf(String id) =>
      folders
          .cast<VaultFolder?>()
          .firstWhere((item) => item!.id == id, orElse: () => null)
          ?.name ??
      documents
          .cast<VaultDocument?>()
          .firstWhere((item) => item!.id == id, orElse: () => null)
          ?.title;

  List<String> ancestorsOf(String id) {
    final result = <String>[];
    var current = parentOf(id);
    while (current != null && !result.contains(current) && result.length < 64) {
      result.add(current);
      current = parentOf(current);
    }
    return result.reversed.toList();
  }

  String pathOf(String id) =>
      [...ancestorsOf(id), id].map(nameOf).whereType<String>().join('/');

  Set<String> descendantsOf(String id) {
    final ids = {id};
    final pending = [id];
    while (pending.isNotEmpty) {
      final parent = pending.removeLast();
      for (final folder in folders.where((item) => item.parentId == parent)) {
        if (ids.add(folder.id)) pending.add(folder.id);
      }
      ids.addAll(documents
          .where((item) => item.parentId == parent)
          .map((item) => item.id));
    }
    return ids;
  }

  static String validName(String value) {
    final name = value.trim();
    if (name.isEmpty ||
        name == '.' ||
        name == '..' ||
        name.length > 240 ||
        name.contains(RegExp(r'[/\\:\n\r]'))) {
      throw VaultException('名称不能为空，不能包含路径分隔符、控制字符或超过 240 字节。');
    }
    return name;
  }

  void _checkParent(String? parent, String toolId) {
    if (!supportedTools.contains(toolId)) throw VaultException('此工具不支持文档库。');
    if (parent != null &&
        !folders.any((item) => item.id == parent && item.toolId == toolId)) {
      throw VaultException('目标文件夹不存在或属于其他工具。');
    }
  }

  bool _available(String name, String toolId, String? parent,
      {String? excluding}) {
    String key(String text) => text.toLowerCase();
    return !folders.any((item) =>
            item.id != excluding &&
            item.toolId == toolId &&
            item.parentId == parent &&
            key(item.name) == key(name)) &&
        !documents.any((item) =>
            item.id != excluding &&
            item.toolId == toolId &&
            item.parentId == parent &&
            key(item.title) == key(name));
  }

  String uniqueName(String proposed, String toolId, String? parent) {
    if (_available(proposed, toolId, parent)) return proposed;
    final dot = proposed.lastIndexOf('.');
    final stem = dot > 0 ? proposed.substring(0, dot) : proposed;
    final ext = dot > 0 ? proposed.substring(dot) : '';
    var index = 2;
    while (true) {
      final name = '$stem ($index)$ext';
      if (_available(name, toolId, parent)) return name;
      index++;
    }
  }

  String createFolder(
      {required String toolId,
      required String name,
      String? parent,
      String? id}) {
    _checkParent(parent, toolId);
    final valid = validName(name);
    if (!_available(valid, toolId, parent))
      throw VaultException('目标位置已有同名文件或文件夹。');
    if (parent != null && ancestorsOf(parent).length >= 63)
      throw VaultException('文件夹不能超过 64 层。');
    final folder = VaultFolder(
        id: id ?? _id(), toolId: toolId, name: valid, parentId: parent);
    folders.add(folder);
    return folder.id;
  }

  String createDocument(
      {required String toolId,
      required String name,
      String content = '',
      String? parent,
      String? id}) {
    _checkParent(parent, toolId);
    final valid = validName(name);
    if (!_available(valid, toolId, parent))
      throw VaultException('目标位置已有同名文件或文件夹。');
    final file = VaultDocument(
        id: id ?? _id(),
        toolId: toolId,
        title: valid,
        content: content,
        parentId: parent);
    documents.add(file);
    return file.id;
  }

  void rename(String id, String value) {
    final toolId = toolOf(id);
    if (toolId == null) throw VaultException('所选项目已不存在。');
    final name = validName(value);
    if (!_available(name, toolId, parentOf(id), excluding: id))
      throw VaultException('目标位置已有同名文件或文件夹。');
    for (final folder in folders.where((item) => item.id == id)) {
      folder.name = name;
      folder.modified = DateTime.now();
    }
    for (final file in documents.where((item) => item.id == id)) {
      file.title = name;
      file.modified = DateTime.now();
    }
  }

  void move(String id, String? parent) {
    final toolId = toolOf(id);
    final name = nameOf(id);
    if (toolId == null || name == null) throw VaultException('所选项目已不存在。');
    _checkParent(parent, toolId);
    if (parent != null && descendantsOf(id).contains(parent))
      throw VaultException('不能将文件夹移动到自身内部。');
    if (!_available(name, toolId, parent, excluding: id))
      throw VaultException('目标位置已有同名文件或文件夹。');
    for (final folder in folders.where((item) => item.id == id)) {
      folder.parentId = parent;
      folder.modified = DateTime.now();
    }
    for (final file in documents.where((item) => item.id == id)) {
      file.parentId = parent;
      file.modified = DateTime.now();
    }
  }

  String duplicate(String id) {
    final file = documents
        .cast<VaultDocument?>()
        .firstWhere((item) => item!.id == id, orElse: () => null);
    if (file == null) throw VaultException('只能复制文档。');
    final copyId = createDocument(
        toolId: file.toolId,
        name: uniqueName(file.title, file.toolId, file.parentId),
        content: file.content,
        parent: file.parentId);
    documents
        .firstWhere((item) => item.id == copyId)
        .metadata
        .addAll(file.metadata);
    return copyId;
  }

  Set<String> delete(String id) {
    final removed = descendantsOf(id);
    documents.removeWhere((item) => removed.contains(item.id));
    folders.removeWhere((item) => removed.contains(item.id));
    return removed;
  }

  List<VaultNode> tree(String toolId,
      {VaultSort sort = VaultSort.name,
      String query = '',
      bool includeContent = false}) {
    List<VaultNode> build(String? parent, String prefix) {
      final folderNodes = folders
          .where((item) => item.toolId == toolId && item.parentId == parent)
          .map((folder) {
        final children = build(
            folder.id, prefix.isEmpty ? folder.name : '$prefix/${folder.name}');
        return VaultNode(
            id: folder.id,
            title: folder.name,
            path: prefix.isEmpty ? folder.name : '$prefix/${folder.name}',
            isFolder: true,
            modified: folder.modified,
            created: folder.created,
            children: children);
      });
      final fileNodes = documents
          .where((item) => item.toolId == toolId && item.parentId == parent)
          .map((file) {
        return VaultNode(
            id: file.id,
            title: file.title,
            path: prefix.isEmpty ? file.title : '$prefix/${file.title}',
            isFolder: false,
            modified: file.modified,
            created: file.created,
            children: const []);
      });
      final nodes = [...folderNodes, ...fileNodes];
      nodes.sort((a, b) {
        if (a.isFolder != b.isFolder) return a.isFolder ? -1 : 1;
        return switch (sort) {
          VaultSort.modified => b.modified.compareTo(a.modified),
          VaultSort.created => b.created.compareTo(a.created),
          VaultSort.name =>
            a.title.toLowerCase().compareTo(b.title.toLowerCase()),
        };
      });
      if (query.trim().isEmpty) return nodes;
      final needle = query.toLowerCase();
      return [
        for (final node in nodes)
          if (node.isFolder)
            VaultNode(
              id: node.id,
              title: node.title,
              path: node.path,
              isFolder: true,
              modified: node.modified,
              created: node.created,
              children: node.children,
            )
          else if (node.title.toLowerCase().contains(needle) ||
              (includeContent &&
                  documents.any((file) =>
                      file.id == node.id &&
                      file.content.toLowerCase().contains(needle))))
            node
      ]
          .where((node) =>
              !node.isFolder ||
              node.children.isNotEmpty ||
              node.title.toLowerCase().contains(needle))
          .toList();
    }

    return build(null, '');
  }

  Map<String, Object?> toJson() => {
        'documents': [for (final item in documents) item.toJson()],
        'folders': [for (final item in folders) item.toJson()],
      };

  factory DocumentVault.fromJson(Map<String, Object?> json) => DocumentVault(
        documents: [
          for (final item in json['documents'] as List? ?? const [])
            if (item is Map)
              VaultDocument.fromJson(Map<String, Object?>.from(item)),
        ],
        folders: [
          for (final item in json['folders'] as List? ?? const [])
            if (item is Map)
              VaultFolder.fromJson(Map<String, Object?>.from(item)),
        ],
      );

  static String _id() =>
      'doc-${DateTime.now().microsecondsSinceEpoch}-${documentsSeed++}';
}

int documentsSeed = 0;

File: next-macos-native/Sources/MooToolNextCore/DocumentVault.swift
import Foundation

public struct EditorViewState: Codable, Equatable, Sendable {
    public var location = 0
    public var length = 0
    public var scrollX = 0.0
    public var scrollY = 0.0
    public init(location: Int = 0, length: Int = 0, scrollX: Double = 0, scrollY: Double = 0) {
        self.location = location; self.length = length; self.scrollX = scrollX; self.scrollY = scrollY
    }
    public func clamped(toUTF16Length count: Int) -> Self {
        let start = max(0, min(location, max(0, count)))
        return Self(location: start, length: max(0, min(length, max(0, count) - start)),
                    scrollX: scrollX.isFinite ? max(0, scrollX) : 0, scrollY: scrollY.isFinite ? max(0, scrollY) : 0)
    }
}

public struct DocumentFolder: Codable, Equatable, Identifiable {
    public var id = UUID()
    public var toolID: String
    public var parentID: UUID?
    public var name: String
    public var created = Date()
    public var modified = Date()
    public init(toolID: String, name: String, parentID: UUID? = nil) { self.toolID = toolID; self.name = name; self.parentID = parentID }
}

public enum VaultSort: String, Codable, CaseIterable { case name, modified, created
    public var title: String { switch self { case .name: return "名称"; case .modified: return "修改时间"; case .created: return "创建时间" } }
}
public enum NoteViewMode: String, Codable, CaseIterable { case editor, split, preview
    public var title: String { switch self { case .editor: return "编辑"; case .split: return "分栏"; case .preview: return "预览" } }
}
public struct VaultPreferences: Codable, Equatable {
    public var selectedEntryID: UUID?
    public var expanded: Set<UUID> = []
    public var query = ""
    public var includeContent = true
    public var sort: VaultSort = .name
    public var treeVisible = true
    public var noteViewMode: NoteViewMode = .editor
    public init() {}
}

public struct VaultNode: Identifiable {
    public let id: UUID
    public let title: String
    public let path: String
    public let isFolder: Bool
    public let modified: Date
    public let created: Date
    public let children: [VaultNode]
}
public struct VaultRow: Identifiable {
    public var id: UUID { node.id }
    public let node: VaultNode
    public let depth: Int
}
public struct DocumentImportItem: Sendable {
    public let relativePath: String
    public let content: String
    public init(relativePath: String, content: String) { self.relativePath = relativePath; self.content = content }
}

/// An independent, ID-based document library. Names and paths never address another product's files.
public struct DocumentVault {
    public var documents: [SavedDocument]
    public var folders: [DocumentFolder]
    public init(documents: [SavedDocument] = [], folders: [DocumentFolder] = []) { self.documents = documents; self.folders = folders }
    public func validate(language: AppLanguage = AppLocalization.preferredLanguage()) throws {
        let allIDs = documents.map(\.id) + folders.map(\.id)
        guard Set(allIDs).count == allIDs.count else { throw Self.err("vault.error.duplicateIDs", language) }
        let byID = Dictionary(uniqueKeysWithValues: folders.map { ($0.id, $0) })
        for folder in folders {
            guard ["json", "quickNote"].contains(folder.toolID) else { throw Self.err("vault.error.invalidFolderTool", language) }
            _ = try Self.validName(folder.name, language: language)
            var visited: Set<UUID> = [folder.id], parent = folder.parentID
            while let id = parent {
                guard let ancestor = byID[id], ancestor.toolID == folder.toolID, visited.insert(id).inserted, visited.count <= 64 else { throw Self.err("vault.error.invalidFolderHierarchy", language) }
                parent = ancestor.parentID
            }
        }
        for file in documents {
            guard ["json", "quickNote"].contains(file.toolID) else { throw Self.err("vault.error.invalidDocumentTool", language) }
            if let parent = file.parentID { guard byID[parent]?.toolID == file.toolID else { throw Self.err("vault.error.documentParentMismatch", language) } }
        }
    }
    public func parent(of id: UUID) -> UUID? { folders.first { $0.id == id }?.parentID ?? documents.first { $0.id == id }?.parentID }
    public func tool(of id: UUID) -> String? { folders.first { $0.id == id }?.toolID ?? documents.first { $0.id == id }?.toolID }
    public func name(of id: UUID) -> String? { folders.first { $0.id == id }?.name ?? documents.first { $0.id == id }?.title }
    public func ancestors(of id: UUID) -> [UUID] {
        var result: [UUID] = [], current = parent(of: id)
        while let value = current, !result.contains(value), result.count < 64 { result.append(value); current = parent(of: value) }
        return result.reversed()
    }
    public func path(of id: UUID) -> String { (ancestors(of: id) + [id]).compactMap(name).joined(separator: "/") }
    public func descendants(of id: UUID) -> Set<UUID> {
        var ids: Set<UUID> = [id], pending = [id]
        while let parent = pending.popLast() {
            for folder in folders where folder.parentID == parent { if ids.insert(folder.id).inserted { pending.append(folder.id) } }
            ids.formUnion(documents.filter { $0.parentID == parent }.map(\.id))
        }; return ids
    }
    public static func validName(_ value: String, language: AppLanguage = AppLocalization.preferredLanguage()) throws -> String {
        let name = value.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !name.isEmpty, name != ".", name != "..", name.utf8.count <= 240,
              !name.contains(where: { $0 == "/" || $0 == "\\" || $0 == ":" || $0.isNewline }),
              !name.unicodeScalars.contains(where: { CharacterSet.controlCharacters.contains($0) }) else { throw err("vault.error.invalidName", language) }
        return name
    }
    private static func err(_ key: String, _ language: AppLanguage) -> ToolError {
        ToolError(AppLocalization.string(key, language: language))
    }
    private func checkParent(_ parent: UUID?, toolID: String, language: AppLanguage) throws {
        guard ["json", "quickNote"].contains(toolID) else { throw Self.err("vault.error.unsupportedTool", language) }
        if let parent { guard folders.contains(where: { $0.id == parent && $0.toolID == toolID }) else { throw Self.err("vault.error.targetFolderMissing", language) } }
    }
    private func available(_ name: String, toolID: String, parent: UUID?, excluding: UUID? = nil) -> Bool {
        func key(_ text: String) -> String { text.precomposedStringWithCanonicalMapping.lowercased() }
        return !folders.contains { $0.id != excluding && $0.toolID == toolID && $0.parentID == parent && key($0.name) == key(name) }
            && !documents.contains { $0.id != excluding && $0.toolID == toolID && $0.parentID == parent && key($0.title) == key(name) }
    }
    public func uniqueName(_ proposed: String, toolID: String, parent: UUID?) -> String {
        if available(proposed, toolID: toolID, parent: parent) { return proposed }
        let url = URL(fileURLWithPath: proposed), ext = url.pathExtension
        let stem = ext.isEmpty ? proposed : String(proposed.dropLast(ext.count + 1))
        var index = 2
        while true {
            let name = "\(stem) (\(index))" + (ext.isEmpty ? "" : "." + ext)
            if available(name, toolID: toolID, parent: parent) { return name }; index += 1
        }
    }
    @discardableResult public mutating func createFolder(toolID: String, name: String, parent: UUID? = nil, language: AppLanguage = AppLocalization.preferredLanguage()) throws -> UUID {
        try checkParent(parent, toolID: toolID, language: language); let name = try Self.validName(name, language: language)
        guard available(name, toolID: toolID, parent: parent) else { throw Self.err("vault.error.nameConflict", language) }
        if let parent { guard ancestors(of: parent).count < 63 else { throw Self.err("vault.error.folderDepthLimit", language) } }
        let folder = DocumentFolder(toolID: toolID, name: name, parentID: parent); folders.append(folder); return folder.id
    }
    @discardableResult public mutating func createDocument(toolID: String, name: String, content: String = "", parent: UUID? = nil, language: AppLanguage = AppLocalization.preferredLanguage()) throws -> UUID {
        try checkParent(parent, toolID: toolID, language: language); let name = try Self.validName(name, language: language)
        guard available(name, toolID: toolID, parent: parent) else { throw Self.err("vault.error.nameConflict", language) }
        var file = SavedDocument(toolID: toolID, title: name, content: content); file.parentID = parent; file.created = Date()
        documents.append(file); return file.id
    }
    public mutating func rename(_ id: UUID, to value: String, language: AppLanguage = AppLocalization.preferredLanguage()) throws {
        guard let toolID = tool(of: id) else { throw Self.err("vault.error.selectionMissing", language) }
        let name = try Self.validName(value, language: language)
        guard available(name, toolID: toolID, parent: parent(of: id), excluding: id) else { throw Self.err("vault.error.nameConflict", language) }
        if let index = folders.firstIndex(where: { $0.id == id }) { folders[index].name = name; folders[index].modified = Date() }
        if let index = documents.firstIndex(where: { $0.id == id }) { documents[index].title = name; documents[index].modified = Date() }
    }
    public mutating func move(_ id: UUID, to parent: UUID?, language: AppLanguage = AppLocalization.preferredLanguage()) throws {
        guard let toolID = tool(of: id), let name = name(of: id) else { throw Self.err("vault.error.selectionMissing", language) }
        try checkParent(parent, toolID: toolID, language: language)
        if let parent, descendants(of: id).contains(parent) { throw Self.err("vault.error.moveIntoSelf", language) }
        guard available(name, toolID: toolID, parent: parent, excluding: id) else { throw Self.err("vault.error.targetNameConflict", language) }
        var copy = self
        if let index = copy.folders.firstIndex(where: { $0.id == id }) { copy.folders[index].parentID = parent; copy.folders[index].modified = Date() }
        if let index = copy.documents.firstIndex(where: { $0.id == id }) { copy.documents[index].parentID = parent; copy.documents[index].modified = Date() }
        try copy.validate(language: language); self = copy
    }
    @discardableResult public mutating func duplicate(_ id: UUID, language: AppLanguage = AppLocalization.preferredLanguage()) throws -> UUID {
        guard var file = documents.first(where: { $0.id == id }) else { throw Self.err("vault.error.selectDocumentToCopy", language) }
        file.id = UUID(); file.title = uniqueName(file.title, toolID: file.toolID, parent: file.parentID)
        file.created = Date(); file.modified = Date(); documents.append(file); return file.id
    }
    @discardableResult public mutating func delete(_ id: UUID) -> Set<UUID> {
        let ids = descendants(of: id); folders.removeAll { ids.contains($0.id) }; documents.removeAll { ids.contains($0.id) }; return ids
    }
    /// Batch operations commit only after every item has been validated.
    public mutating func importDocuments(_ items: [DocumentImportItem], toolID: String, parent: UUID? = nil, language: AppLanguage = AppLocalization.preferredLanguage()) throws -> [UUID] {
        try checkParent(parent, toolID: toolID, language: language)
        guard items.count <= 500, items.reduce(0, { $0 + $1.content.utf8.count }) <= 32 * 1024 * 1024 else { throw Self.err("vault.import.error.limits", language) }
        var copy = self, imported: [UUID] = []
        for item in items {
            let parts = item.relativePath.components(separatedBy: "/")
            guard !parts.isEmpty, parts.count <= 64, item.content.utf8.count <= 10 * 1024 * 1024 else { throw Self.err("vault.error.importPathLimits", language) }
            for part in parts { _ = try Self.validName(part, language: language) }
            var directory = parent
            for part in parts.dropLast() {
                if let existing = copy.folders.first(where: { $0.toolID == toolID && $0.parentID == directory && $0.name.precomposedStringWithCanonicalMapping.lowercased() == part.precomposedStringWithCanonicalMapping.lowercased() }) { directory = existing.id }
                else { directory = try copy.createFolder(toolID: toolID, name: part, parent: directory, language: language) }
            }
            let name = copy.uniqueName(parts.last!, toolID: toolID, parent: directory)
            imported.append(try copy.createDocument(toolID: toolID, name: name, content: item.content, parent: directory, language: language))
        }
        try copy.validate(language: language); self = copy; return imported
    }
    public func tree(toolID: String, preferences: VaultPreferences) -> [VaultNode] {
        let query = preferences.query.trimmingCharacters(in: .whitespacesAndNewlines)
        let files = Dictionary(grouping: documents.filter { $0.toolID == toolID }, by: \.parentID)
        let directories = Dictionary(grouping: folders.filter { $0.toolID == toolID }, by: \.parentID)
        func build(parent: UUID?, prefix: String, includeAll: Bool, depth: Int) -> [VaultNode] {
            guard depth <= 64 else { return [] }
            var nodes: [VaultNode] = []
            for folder in directories[parent] ?? [] {
                let path = prefix + folder.name, matches = includeAll || folder.name.localizedCaseInsensitiveContains(query)
                let children = build(parent: folder.id, prefix: path + "/", includeAll: matches, depth: depth + 1)
                if matches || !children.isEmpty { nodes.append(VaultNode(id: folder.id, title: folder.name, path: path, isFolder: true, modified: folder.modified, created: folder.created, children: children)) }
            }
            for file in files[parent] ?? [] where includeAll || file.title.localizedCaseInsensitiveContains(query) || (preferences.includeContent && file.content.localizedCaseInsensitiveContains(query)) {
                nodes.append(VaultNode(id: file.id, title: file.title, path: prefix + file.title, isFolder: false, modified: file.modified, created: file.created ?? file.modified, children: []))
            }
            return nodes.sorted {
                if $0.isFolder != $1.isFolder { return $0.isFolder }
                if preferences.sort == .modified && $0.modified != $1.modified { return $0.modified > $1.modified }
                if preferences.sort == .created && $0.created != $1.created { return $0.created > $1.created }
                let order = $0.title.localizedStandardCompare($1.title)
                return order == .orderedSame ? $0.id.uuidString < $1.id.uuidString : order == .orderedAscending
            }
        }
        return build(parent: nil, prefix: "", includeAll: query.isEmpty, depth: 0)
    }
    public func rows(toolID: String, preferences: VaultPreferences) -> [VaultRow] {
        func flatten(_ nodes: [VaultNode], depth: Int) -> [VaultRow] {
            nodes.flatMap { node in
                [VaultRow(node: node, depth: depth)] + (node.isFolder && (preferences.expanded.contains(node.id) || !preferences.query.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty) ? flatten(node.children, depth: depth + 1) : [])
            }
        }; return flatten(tree(toolID: toolID, preferences: preferences), depth: 0)
    }
}
